### PR TITLE
Make WriteAsync wait for changes asynchronously

### LINKED ISSUE #1728: WriteAsync blocks on Refresh

### PR DESCRIPTION
**Pull request by fealebenpae**
## Description

Fixes #1728

##  TODO

* [x] Changelog entry
* [x] Tests (if applicable)
* [x] Update PCL (if applicable)

### CODE REVIEW

File: Realm/Realm/Realm.cs [hunk 1]
```diff
@@ -695,13 +695,23 @@ public Task WriteAsync(Action<Realm> action)
...
                 {
                     using (var realm = GetInstance(Config))
                     {
                         realm.Write(() => action(realm));
+                        otherVersion = realm.SharedRealmHandle.GetTransactionVersion();
                     }
-                }).ContinueWith(_ => Refresh(), scheduler);
+                }).ContinueWith(_ =>
+                {
+                    if (otherVersion == SharedRealmHandle.GetTransactionVersion())
```
Comment (nirinchev) [inline Realm/Realm/Realm.cs]: Should this be strict equality? I imagine if the current version is higher (for some reason) than the other version, we should still return?

File: Realm/Realm/Realm.cs [hunk 2]
```diff
@@ -724,6 +734,30 @@ public bool Refresh()
...
+            {
+                Refresh();
+                return Task.CompletedTask;
+            }
+
+            var tcs = new TaskCompletionSource<object>();
+            RealmChangedEventHandler handler = null;
+            handler = (sender, e) =>
+            {
+                ((Realm)sender).RealmChanged -= handler;
```
Comment (nirinchev) [inline Realm/Realm/Realm.cs]: I'm a bit hazy on the subtleties of doing it like this (via the `sender`) rather than unsubscribing `this.RealmChanged`. Is it to avoid capturing `this` in the closure and what would the harm be if we did capture it?

Comment (fealebenpae) [reply Realm/Realm/Realm.cs]: Exactly - I didn’t want to capture `this` and create a reference cycle. If we capture it and the wvent never fires then the realm will leak.

File: Realm/Realm/Realm.cs [hunk 3]
```diff
@@ -724,6 +734,30 @@ public bool Refresh()
             return SharedRealmHandle.Refresh();
         }
 
+        /// <summary>
+        /// Waits for the realm and outstanding objects to be updated to a newer version.
+        /// </summary>
+        /// <returns>An awaitable <see cref="Task"/>.</returns>
+        private Task WaitForChangesAsync()
```
Comment (nielsenko) [inline Realm/Realm/Realm.cs]: I hope you don't mind a comment from an outsider. Could you consider making this public? And why the change in name compared to `Refresh`? I would expect it to be called `RefreshAsync`

Comment (nielsenko) [reply Realm/Realm/Realm.cs]: Hmm.. I think I get it. You need to compare the current transaction version number with something, to avoid waiting in vane for a change that may never come. Is it not possible to query the realm for the highest transaction version seen?

**Comment by nirinchev** [comment]
@tgoyne I added you as a reviewer to give your opinion on the high-level approach and confirm whether it matches what we discussed, thanks.

**Comment by nirinchev** [comment]
Here are some packages:
https://drive.google.com/file/d/1bQ7U5jBzg4xSvcbkaxNObn1s8JIGvVrV/view?usp=sharing
https://drive.google.com/file/d/19b0mie_oHmQbZb2imje2XJhxUZqmyjEC/view?usp=sharing
